File: hello-world/rest-api/cmx-hello-world.py
# Hello World example of calling the Custom Metrics REST API using Python

import time
import random
import requests

# Host of port where Aternity APM agent is install. This example assumes the
# agent is local. If not, replace localhost with the agent's hostname and 
# ensure the agent is accessible to remote clients.
HOST = "localhost"  # Hostname of where 
PORT = 7074
BASE_URL = 'https://%s:%s/Aternity/CustomMetrics/1.0.0' % (HOST, PORT)


def helloWorldPython():
    """
    Add a random number every minute for five minutes
    """

    # First, you'll need to define a metric. This is done by hitting the 
    # "/metric" endpoint. The following example defines a metric with the id 
    # "hello-world-python". It also provides a display name, description and
    # unit for the metric.
    metric_definition = {
        "version": 1,
        "metric-id": "hello-world-python",
        "display-name": "Hello World Python",
        "description": "Metric from Hello World Python Example",
        "units": "count"
    }

    # POST the metric definition to the Aternity APM agent
    requests.post(BASE_URL + "/metric", json={"metric-definitions": [metric_definition]}, verify=False)

    for i in range(5):

        timestamp = int(time.time())
        value = random.randint(1,10)

        # Next, you'll want to publish metric values (samples) by hitting the 
        # "/samples" endpoint. The following example publishes a single metric
        # sample to the "hello-world-python" metric previously defined. The 
        # value is a random number and the timestamp is the current timestamp 
        # in seconds. It also sets a single dimension: "myDimension:test-1".
        sample = {
            "metric-id": "hello-world-python",
            "source": "python",
            "timestamp": [timestamp],
            "value": [value],
            "dimensions":{
                "myDimension": "test-1"
            },
            "tags":{}
        }

        # PUT the metric sample to the Aternity APM agent
        requests.put(BASE_URL + "/samples", json={"metric-samples": [sample]}, verify=False)

        print("  - Metric sample published. Sleeping 60s...")
        
        # Sleep for 60 seconds
        time.sleep(60)

    return

if __name__ == "__main__":
    print("Starting Custom Metrics Hello World Python Example...")
    helloWorldPython()

File: hello-world/java-sdk/src/main/java/com/aternity/samples/CmxHelloWorld.java
package com.aternity.samples;

import java.util.HashMap;
import java.util.Map;
import java.util.Random;

import com.riverbed.agent.cmxsdk.api.CmxClientEnvironment.CmxClientEnvironmentBuilder;
import com.riverbed.agent.cmxsdk.api.CmxMetricDefinition.CmxMetricDefinitionBuilder;
import com.riverbed.agent.cmxsdk.api.CmxSdkClient;
import com.riverbed.agent.cmxsdk.api.CmxSdkException;
import com.riverbed.agent.cmxsdk.api.ICmxClientEnvironment;
import com.riverbed.agent.cmxsdk.api.ICmxMetric;
import com.riverbed.agent.cmxsdk.api.ICmxMetricDefinition;

/**
 * Hello World example of calling the Custom Metrics Java SDK API.
 * 
 * Note: the following example assumes the Java application is being run
 * locally on a system with the Aternity APM Agent installed.
 */
public class CmxHelloWorld {

    static final Random random = new Random();

    public static void main(String[] args) {

        System.out.println("Starting Custom Metrics Hello World Java Example...");
        CmxSdkClient client = null;
        
        try {
            // Open a connection to the Aternity APM Agent (cmx server component)        
        	ICmxClientEnvironment env = CmxClientEnvironmentBuilder.defaultBuilder("hello-world-java").build();		
			client = new CmxSdkClient(env);
			client.open();
			
			 // Define a set of dimensions. The following example sets a single dimension: "myDimension:test-1".
            Map<String, String> dimensions = new HashMap<String, String>();
            dimensions.put("myDimension", "test-1");
                       
            // Define a metric. The following example defines a metric with the id "hello-world-java". 
            // It also provides a display name, description and unit for the metric.
            ICmxMetricDefinition helloWorldJavaMetricDef = new CmxMetricDefinitionBuilder("hello-world-java")
                    .version(1)
                    .description("Metric from Hello World Java Example")
                    .displayName("Hello World Java")
                    .units("count")
                    .build();

            ICmxMetric helloWorldJavaMetric = client.createMetric(helloWorldJavaMetricDef, dimensions);
            
            for (int ctr=0; ctr < 5; ctr++) {
                // Next, publish metric values (samples). The following example publishes a single metric sample to the 
                // "hello-world-java" metric defined previously. The value is a random number between 0 and 100 and the
                // timestamp is the current timestamp in seconds.
                long currTime = System.currentTimeMillis() / 1000;
                helloWorldJavaMetric.addSample(currTime, random.nextInt(100));

                // Sleep 5s before publishing another metric sample
                System.out.println("  - Metric sample published. Sleeping 5s...");
                Thread.sleep(5000);
            }
          
        } catch (Exception e) {
            e.printStackTrace();

        } finally {
            // Close the connection to the Aternity APM Agent (cmx server component)
            if (client != null) {
                try {
                    client.close();
                } catch (CmxSdkException e) {
                    e.printStackTrace();
                }
            }
        }
    }
}